### ISSUE TITLE
[BUG] CryptographicException: Padding is invalid and cannot be removed.

### ISSUE THREAD

**Issue by bsumter** [Azure/azure-sdk-for-net#25656]
### Library name and version

Azure.Storage.Blobs 12.10.0

### Describe the bug

An exception is thrown when doing multiple UploadAsync and DownloadAsync in parallel with ClientSideEncryption enabled.

Exception:
`System.Security.Cryptography.CryptographicException: Padding is invalid and cannot be removed when doing`

Background information:
We are migrating from Microsoft.Azure.Storage.Blob to Azure.Storage.Blobs. While doing a stress test on our service, we started seeing these exceptions. I've isolated the code to a single Unit Test which should be helpful to reproduce the issue. I've also tried several variants resulting in the same exception:
 - New BlobClient per request
 - Using streams for uploading and downloading

### Expected behavior

No exception.

### Actual behavior

A `System.Security.Cryptography.CryptographicException: Padding is invalid and cannot be removed when doing` is thrown.

### Reproduction Steps

```C#
[TestMethod]
public async Task TestUploadAsyncScenarios2()
{
    var keyVaultUri = new Uri("https://***.vault.azure.net");
    var keyVaultSecretName = "***";
    var keyVaultTenantId = "***";
    var keyVaultClientId = "***";
    var certificateThumbprint = "***";

    var x509Store = new X509Store(StoreName.My, StoreLocation.CurrentUser);
    x509Store.Open(OpenFlags.ReadOnly);
    var certificates = x509Store.Certificates.Find(X509FindType.FindByThumbprint, certificateThumbprint, false);
    var certificate = certificates[0];
    x509Store.Close();

    var credential = new ClientCertificateCredential(keyVaultTenantId, keyVaultClientId, certificate);
    var secretClient = new SecretClient(keyVaultUri, credential);
    var secret = secretClient.GetSecret(keyVaultSecretName);

    var keyResolver = new KeyResolver(credential);
    var keyEncryptionKey = keyResolver.Resolve(secret.Value.Id);

    var clientSideEncryptionOptions = new ClientSideEncryptionOptions(ClientSideEncryptionVersion.V1_0)
    {
        KeyEncryptionKey = keyEncryptionKey,
        KeyResolver = keyResolver,
        KeyWrapAlgorithm = "A256KW"
    };

    var blobConnectionString = "***";
    var blobContainerName = "***";
    var blobName = "test";
    var blobContents = "test content";

    var blobContainerClient = new BlobContainerClient(blobConnectionString, blobContainerName);
    var blobClient = blobContainerClient.GetBlobClient(blobName);
    var encryptedBlobClient = blobClient.WithClientSideEncryptionOptions(clientSideEncryptionOptions);

    await UploadAsync();

    while (true)
    {
        try
        {
            await Task.WhenAll(
                UploadAsync(),
                DownloadAsync(),
                UploadAsync()
            );
        }
        catch (Exception)
        {
            throw;
        }
    }

    Task UploadAsync() => encryptedBlobClient.UploadAsync(new BinaryData(blobContents), new BlobUploadOptions { });
    Task DownloadAsync() => encryptedBlobClient.DownloadContentAsync();
}
```

### Environment

Windows 10
.NET Standard 2.0
Visual Studio 16.10.3

**Comment by SaurabhSharma-MSFT** [comment]
@bsumter We are looking into this and get back to you in case need any information.

**Comment by bsumter** [comment]
As a side note, we see a significant performance decrease since the upgrade from `Microsoft.Azure.Storage.Blob` to `Azure.Storage.Blobs`.

**Comment by MughundhanRaveendran-MSFT** [comment]
@bsumter ,

For the MyKeyProvider.GetKey(), and MyKeyProvider.GetKeyResolver(), is it through KeyVault SDK or an external implementation?

https://docs.microsoft.com/en-us/azure/storage/common/storage-client-side-encryption?tabs=dotnet#blob-service-encryption

Please provide information about this class

**Comment by bsumter** [comment]
Using KeyVault SDK. I've updated the initial post including the KeyProvider.

**Comment by MughundhanRaveendran-MSFT** [comment]
@bsumter , Sorry to get back and forth. I just wanted to repro this issue with the same code as yours. The X509StoreAdapter and X509Certificate2Retriever arent a part of the System.Security.Cryptography.X509Certificates namespace. I have checked it for .net standard 2.0 and latest .net core versions (.net 3.1,5,6)

https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates?view=netstandard-2.0

Please point out if Im missing out something

**Comment by bsumter** [comment]
> @bsumter , Sorry to get back and forth. I just wanted to repro this issue with the same code as yours. The X509StoreAdapter and X509Certificate2Retriever arent a part of the System.Security.Cryptography.X509Certificates namespace. I have checked it for .net standard 2.0 and latest .net core versions (.net 3.1,5,6)
> 
> https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates?view=netstandard-2.0
> 
> Please point out if Im missing out something

My Bad. I've updated the initial post. Should be using `System.Security.Cryptography.X509Certificates` now.

**Comment by MughundhanRaveendran-MSFT** [comment]
@bsumter , Sorry for the delay in getting back to you.

    var keyVaultSecretName = "***";
    var keyVaultTenantId = "***";
    var keyVaultClientId = "***";
    var certificateThumbprint = "***";

I was facing some trouble running the code. I have a 509 certificate in the keyvault so I can get the thumbprint. For the keyVaultTenantId , keyVaultClientId , keyVaultSecretName - Are you using the service principal's tenant id, client id and secrets which is registered in AAD with Azure keyvault as delegated permissions? or is it something else?

**Comment by bsumter** [comment]
> @bsumter , Sorry for the delay in getting back to you.
> 
> ```
> var keyVaultSecretName = "***";
> var keyVaultTenantId = "***";
> var keyVaultClientId = "***";
> var certificateThumbprint = "***";
> ```
> 
> I was facing some trouble running the code. I have a 509 certificate in the keyvault so I can get the thumbprint. For the keyVaultTenantId , keyVaultClientId , keyVaultSecretName - Are you using the service principal's tenant id, client id and secrets which is registered in AAD with Azure keyvault as delegated permissions? or is it something else?

Sorry, I missed your message. Yes we're using Azure AD.

**Comment by jaschrep-msft** [comment]
@bsumter unless I'm sorely misreading your code, I would expect this sample to fail.
```csharp
Task UploadAsync() => encryptedBlobClient.UploadAsync(new BinaryData(blobContents), new BlobUploadOptions { });
Task DownloadAsync() => encryptedBlobClient.DownloadContentAsync();

await UploadAsync();

await Task.WhenAll(
    UploadAsync(),
    DownloadAsync(),
    UploadAsync()
);
```

If any of those uploads in the `WhenAll` finish before the download then your download contents will have shifted out from under you with a different content encryption key and you'll be expected to fail no matter what. The padding error is the result of trying to decrypt invalid contents, which your sample here is creating.

All uploads use a new, securely generated content encryption key, while the key you provide with keyvault is a key encryption key. You can look into the storage library [documentation](https://docs.microsoft.com/en-us/azure/storage/common/storage-client-side-encryption?tabs=dotnet#encryption-and-decryption-via-the-envelope-technique) on this for more information.

> As a side note: we see a significant performance decrease since the upgrade from `Microsoft.Azure.Storage.Blob` to `Azure.Storage.Blobs`.

I recommend this [blog post](https://devblogs.microsoft.com/azure-sdk/tuning-your-uploads-and-downloads-with-the-azure-storage-client-library-for-net/) for performance concerns.

**Comment by bsumter** [comment]
@jaschrep-msft

Thank you for taking the time to look into this and sharing your knowledge.

> @bsumter unless I'm sorely misreading your code, I would expect this sample to fail.

Are you then also saying this code should be failing when using a similar setup with the deprecated package `Microsoft.Azure.Storage.Blob`? I've tried this, and that code runs perfectly fine. Meaning there's a fundamental difference between the two packages in terms of behavior.

> I recommend this [blog post](https://devblogs.microsoft.com/azure-sdk/tuning-your-uploads-and-downloads-with-the-azure-storage-client-library-for-net/) for performance concerns.

We are working with blobs ranging from 10KB to 2MB.  Which settings do you recommend us looking into?

**Comment by jaschrep-msft** [comment]
> Are you then also saying this code should be failing when using a similar setup with the deprecated package `Microsoft.Azure.Storage.Blob`? I've tried this, and that code runs perfectly fine. Meaning there's a fundamental difference between the two packages in terms of behavior.

@bsumter  Your code sample, specifically `await Task.WhenAll(UploadAsync(), DownloadAsync(), UploadAsync());`, is creating a race condition. Given the entirely different internals of Microsoft.Azure.Storage.Blob vs Azure.Storage.Blobs, I can imagine race conditions might be more likely to play out one way vs another in a given implementation. I could not tell you exactly why you are seeing the difference, but I can tell you that uploading to and downloading from the same exact blob concurrently will produce undesirable results, and that is what your posted code does. I would suggest altering your test to use separate blobs when stressing it with simultaneous transfers.

> We are working with blobs ranging from 10KB to 2MB. Which settings do you recommend us looking into?

Hmm. The linked blog post is more to do with files in the 10s of MB and larger, depending on connection speed. To keep this issue on topic, can you open a separate issue with more advanced numbers on the kind of performance degradation you're seeing? Code samples would be helpful for such an issue, including the SDK methods you use to acquire your clients, not just what you do with them once you have them.
